File: backend/scrape-latest-jobs.js
/**
 * Catch.co.kr 최신 채용공고 30건 스크래핑 (JavaScript/Puppeteer)
 * 사용법: node scrape-latest-jobs.js
 * 또는 함수 import: import { scrapeLatestJobs } from './scrape-latest-jobs.js'
 */

import puppeteer from 'puppeteer';
import mysql from 'mysql2/promise';
import dotenv from 'dotenv';

dotenv.config();

const DB_CONFIG = {
  host: process.env.DB_HOST || 'localhost',
  user: process.env.DB_USER || 'root',
  password: process.env.DB_PASSWORD || '',
  database: process.env.DB_NAME || 'commitjob',
  port: process.env.DB_PORT || 3306
};

const BASE_URL = 'https://www.catch.co.kr/';

async function scrapeLatestJobs(pool = null) {
  let browser = null;
  let connection = null;
  let shouldCloseConnection = false;

  const stats = {
    total_scraped: 0,
    total_inserted: 0,
    duplicates: 0,
    errors: 0
  };

  try {
    console.log(`\n${'='.repeat(60)}`);
    console.log('🚀 Catch.co.kr 채용공고 스크래핑 시작');
    console.log('   목표: 최신 채용공고 30건');
    console.log('='.repeat(60) + '\n');

    // MySQL 연결
    if (pool) {
      connection = pool;
      console.log('✅ 기존 DB pool 사용\n');
    } else {
      connection = await mysql.createConnection(DB_CONFIG);
      shouldCloseConnection = true;
      console.log('✅ MySQL 연결 성공\n');
    }

    // Puppeteer 브라우저 실행
    browser = await puppeteer.launch({
      headless: 'new',
      args: [
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--disable-dev-shm-usage',
        '--disable-gpu'
      ]
    });

    const page = await browser.newPage();
    await page.setViewport({ width: 1920, height: 1080 });
    page.setDefaultNavigationTimeout(60000);

    // 이미지 로딩 비활성화 (성능 향상)
    await page.setRequestInterception(true);
    page.on('request', (req) => {
      if (req.resourceType() === 'image') {
        req.abort();
      } else {
        req.continue();
      }
    });

    console.log('✅ Chrome 드라이버 초기화 성공\n');

    // 채용 검색 페이지로 이동
    console.log('📍 채용 검색 페이지로 이동 중...');
    await page.goto(`${BASE_URL}NCS/RecruitSearch`, { waitUntil: 'domcontentloaded' });
    await new Promise(r => setTimeout(r, 2000));
    console.log('✅ 채용 검색 페이지 진입 성공\n');

    // IT개발 카테고리 스크래핑 (최대 2페이지)
    await scrapeCategoryJobs(page, connection, 'IT개발', 'IT', 2, stats);

    // 30건 미만이면 빅데이터·AI 카테고리도 스크래핑
    if (stats.total_scraped < 30) {
      console.log(`\n📊 남은 목표: ${30 - stats.total_scraped}건\n`);
      // 채용 검색 페이지로 다시 이동
      await page.goto(`${BASE_URL}NCS/RecruitSearch`, { waitUntil: 'domcontentloaded' });
      await new Promise(r => setTimeout(r, 2000));
      await scrapeCategoryJobs(page, connection, '빅데이터·AI', 'BIGDATA_AI', 2, stats);
    }

    // 최종 통계
    console.log(`\n${'='.repeat(60)}`);
    console.log('📊 최종 통계');
    console.log('='.repeat(60));
    console.log(`총 수집: ${stats.total_scraped}건`);
    console.log(`신규 저장: ${stats.total_inserted}건`);
    console.log(`중복 제외: ${stats.duplicates}건`);
    console.log(`오류: ${stats.errors}건`);
    console.log('='.repeat(60) + '\n');

  } catch (error) {
    console.error('❌ 스크래핑 오류:', error);
    throw error;
  } finally {
    if (browser) {
      await browser.close();
      console.log('✅ 리소스 정리 완료\n');
    }
    if (connection && shouldCloseConnection) {
      await connection.end();
    }
  }

  return stats;
}

async function scrapeCategoryJobs(page, connection, categoryName, categoryCode, maxPages, stats) {
  console.log(`\n${'='.repeat(60)}`);
  console.log(`🔍 ${categoryName} 카테고리 스크래핑 시작`);
  console.log('='.repeat(60) + '\n');

  try {
    // 카테고리 선택
    console.log(`📂 카테고리 선택: ${categoryName}`);

    // 직무 버튼 클릭 (XPath 사용)
    await page.waitForFunction(() => {
      const buttons = Array.from(document.querySelectorAll('button.bt'));
      return buttons.some(btn => btn.textContent.includes('직무'));
    }, { timeout: 10000 });

    await page.evaluate(() => {
      const buttons = Array.from(document.querySelectorAll('button.bt'));
      const jobBtn = buttons.find(btn => btn.textContent.includes('직무'));
      if (jobBtn) jobBtn.click();
    });
    await new Promise(r => setTimeout(r, 1000));

    // 카테고리 선택
    const categoryText = categoryCode === 'IT' ? 'IT개발' : '빅데이터·AI';

    await page.waitForFunction((text) => {
      const buttons = Array.from(document.querySelectorAll('button.bt'));
      return buttons.some(btn => btn.textContent.includes(text));
    }, { timeout: 10000 }, categoryText);

    await page.evaluate((text) => {
      const buttons = Array.from(document.querySelectorAll('button.bt'));
      const categoryBtn = buttons.find(btn => btn.textContent.includes(text));
      if (categoryBtn) categoryBtn.click();
    }, categoryText);
    await new Promise(r => setTimeout(r, 2000));
    console.log(`✅ ${categoryName} 카테고리 선택 완료\n`);

    let pageNum = 1;
    let categoryScraped = 0;

    while (pageNum <= maxPages && stats.total_scraped < 30) {
      console.log(`📄 페이지 ${pageNum} 스크래핑 중...`);

      // 페이지 로딩 대기
      await new Promise(r => setTimeout(r, 2000));

      // 채용공고 데이터 추출
      const jobs = await page.$$eval('tbody tr', (rows, category) => {
        return rows.map(row => {
          try {
            const companyElem = row.querySelector('td:nth-child(1) a');
            const titleElem = row.querySelector('td:nth-child(2) a');
            const jobInfoElems = row.querySelectorAll('td:nth-child(3) p');
            const conditionsElems = row.querySelectorAll('td:nth-child(4) p');
            const regInfoElems = row.querySelectorAll('td:nth-child(5) p');

            const company = companyElem?.textContent.trim();
            const title = titleElem?.textContent.trim();
            const url = titleElem?.href;

            const job_info = Array.from(jobInfoElems).map(el => el.textContent.trim()).filter(t => t);
            const conditions = Array.from(conditionsElems).map(el => el.textContent.trim()).filter(t => t);
            const registration_info = Array.from(regInfoElems).map(el => el.textContent.trim()).filter(t => t);

            if (company && title && url) {
              return {
                company,
                title,
                url,
                job_info,
                conditions,
                registration_info,
                category
              };
            }
            return null;
          } catch (e) {
            return null;
          }
        }).filter(job => job !== null);
      }, categoryCode);

      if (jobs.length === 0) {
        console.log('  ⚠️ 데이터 없음\n');
        break;
      }

      // DB에 저장
      const inserted = await insertJobs(connection, jobs, pageNum, stats);
      categoryScraped += jobs.length;
      stats.total_scraped += jobs.length;

      console.log(`  ✅ ${jobs.length}건 수집, ${inserted}건 저장 (중복: ${jobs.length - inserted}건)`);
      console.log(`  📊 전체 진행: ${stats.total_scraped}/30건\n`);

      // 30건 도달 시 중단
      if (stats.total_scraped >= 30) {
        console.log('\n✅ 목표 30건 도달! 스크래핑 종료\n');
        break;
      }

      // 다음 페이지로 이동
      try {
        const nextBtn = await page.$('p.page3 a.ico.next');
        if (nextBtn) {
          await nextBtn.click();
          await new Promise(r => setTimeout(r, 2000));
          pageNum++;
        } else {
          console.log('  ⏭️ 다음 페이지 없음\n');
          break;
        }
      } catch (e) {
        console.log('  ⏭️ 다음 페이지 없음\n');
        break;
      }
    }

    console.log(`\n${categoryName} 카테고리: ${categoryScraped}건 수집 완료\n`);

  } catch (error) {
    console.error(`❌ ${categoryName} 카테고리 스크래핑 실패:`, error.message);
  }
}

async function insertJobs(connection, jobs, pageNum, stats) {
  if (!jobs || jobs.length === 0) return 0;

  let inserted = 0;

  for (const job of jobs) {
    try {
      // 중복 체크
      const [existing] = await connection.execute(
        'SELECT id FROM jobs WHERE url = ? AND category = ?',
        [job.url, job.category]
      );

      if (existing.length > 0) {
        stats.duplicates++;
        continue;
      }

      // 회사명 정규화 (검색용)
      const company_search_key = job.company.toLowerCase().replace(/\s/g, '').replace(/\u00A0/g, '');

      // INSERT
      await connection.execute(
        `INSERT INTO jobs
        (company, title, url, category, page, job_info, conditions, registration_info, scraped_at, company_search_key)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, NOW(), ?)`,
        [
          job.company,
          job.title,
          job.url,
          job.category,
          pageNum,
          JSON.stringify(job.job_info),
          JSON.stringify(job.conditions),
          JSON.stringify(job.registration_info),
          company_search_key
        ]
      );

      inserted++;
      stats.total_inserted++;

    } catch (error) {
      stats.errors++;
      console.error(`  ⚠️ DB 저장 실패: ${error.message}`);
      continue;
    }
  }

  return inserted;
}

// Export function for use as module
export { scrapeLatestJobs };

// 커맨드 라인에서 직접 실행된 경우에만 실행
if (import.meta.url === `file://${process.argv[1]}`) {
  scrapeLatestJobs()
    .then((stats) => {
      console.log('✅ 스크래핑 완료');
      process.exit(0);
    })
    .catch((error) => {
      console.error('실행 실패:', error);
      process.exit(1);
    });
}

File: backend_251018/catch-scraper-service/scrape_latest_jobs.py
"""
Catch.co.kr 최신 채용공고 1000건 스크래핑 -> AWS RDS MySQL
"""

import sys
import io
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager
import pymysql
import time
import json
from datetime import datetime
import traceback

DB_CONFIG = {
    'host': 'database-1.czcg4o8cytan.ap-northeast-2.rds.amazonaws.com',
    'port': 3306,
    'user': 'appuser',
    'password': 'Woolim114!',
    'database': 'appdb',
    'charset': 'utf8mb4'
}

BASE_URL = 'https://www.catch.co.kr/'

class JobScraper:
    def __init__(self):
        self.driver = None
        self.connection = None
        self.stats = {
            'total_scraped': 0,
            'total_inserted': 0,
            'duplicates': 0,
            'errors': 0
        }

    def init_driver(self):
        """Chrome 드라이버 초기화"""
        try:
            chrome_options = Options()
            chrome_options.add_argument('--headless=new')
            chrome_options.add_argument('--no-sandbox')
            chrome_options.add_argument('--disable-dev-shm-usage')
            chrome_options.add_argument('--disable-gpu')
            chrome_options.add_argument('--window-size=1920,1080')
            chrome_options.add_argument('--blink-settings=imagesEnabled=false')

            prefs = {'profile.default_content_setting_values': {'images': 2}}
            chrome_options.add_experimental_option('prefs', prefs)
            chrome_options.add_experimental_option('excludeSwitches', ['enable-logging'])

            service = Service(ChromeDriverManager().install())
            self.driver = webdriver.Chrome(service=service, options=chrome_options)
            self.driver.set_page_load_timeout(30)
            self.driver.implicitly_wait(5)

            print("✅ Chrome 드라이버 초기화 성공\n")
            return True
        except Exception as e:
            print(f"❌ Chrome 드라이버 초기화 실패: {e}")
            return False

    def connect_db(self):
        """DB 연결"""
        try:
            self.connection = pymysql.connect(**DB_CONFIG)
            print("✅ AWS RDS MySQL 연결 성공\n")
            return True
        except Exception as e:
            print(f"❌ DB 연결 실패: {e}")
            return False

    def close(self):
        """리소스 정리"""
        if self.driver:
            self.driver.quit()
        if self.connection:
            self.connection.close()
        print("\n✅ 리소스 정리 완료")

    def navigate_to_recruit_page(self):
        """채용 검색 페이지로 이동"""
        try:
            print("📍 채용 검색 페이지로 이동 중...")
            self.driver.get(BASE_URL)
            time.sleep(2)

            # 채용정보 메뉴 클릭
            recruit_menu = WebDriverWait(self.driver, 10).until(
                EC.element_to_be_clickable((By.XPATH, "//a[@href='/NCS/RecruitSearch']"))
            )
            recruit_menu.click()
            time.sleep(2)
            print("✅ 채용 검색 페이지 진입 성공\n")
            return True
        except Exception as e:
            print(f"❌ 페이지 이동 실패: {e}")
            return False

    def select_category(self, category_name):
        """카테고리 선택 (IT개발 또는 빅데이터·AI)"""
        try:
            print(f"📂 카테고리 선택: {category_name}")

            # 직무 버튼 클릭
            job_category_btn = WebDriverWait(self.driver, 10).until(
                EC.element_to_be_clickable((By.XPATH, "//button[contains(@class, 'bt') and contains(text(), '직무')]"))
            )
            job_category_btn.click()
            time.sleep(1)

            # 카테고리 선택
            if category_name == "IT":
                xpath = "//button[contains(@class, 'bt')]//span[contains(text(), 'IT개발')]/.."
            else:  # BIGDATA_AI
                xpath = "//button[contains(@class, 'bt')]//span[contains(text(), '빅데이터·AI')]/.."

            category_btn = WebDriverWait(self.driver, 10).until(
                EC.element_to_be_clickable((By.XPATH, xpath))
            )
            category_btn.click()
            time.sleep(2)
            print(f"✅ {category_name} 카테고리 선택 완료\n")
            return True
        except Exception as e:
            print(f"❌ 카테고리 선택 실패: {e}")
            return False

    def scrape_page_jobs(self, page_num, category):
        """한 페이지의 채용공고 스크래핑"""
        jobs = []
        try:
            time.sleep(2)

            # 테이블 행 가져오기
            rows = self.driver.find_elements(By.XPATH, "//tbody//tr")

            for row in rows:
                try:
                    # 회사명
                    company = row.find_element(By.XPATH, ".//td[1]//a").text.strip()

                    # 제목 및 URL
                    title_elem = row.find_element(By.XPATH, ".//td[2]//a")
                    title = title_elem.text.strip()
                    url = title_elem.get_attribute('href')

                    # 직무정보
                    job_info_elems = row.find_elements(By.XPATH, ".//td[3]//p")
                    job_info = [elem.text.strip() for elem in job_info_elems if elem.text.strip()]

                    # 조건
                    conditions_elems = row.find_elements(By.XPATH, ".//td[4]//p")
                    conditions = [elem.text.strip() for elem in conditions_elems if elem.text.strip()]

                    # 등록정보
                    reg_info_elems = row.find_elements(By.XPATH, ".//td[5]//p")
                    registration_info = [elem.text.strip() for elem in reg_info_elems if elem.text.strip()]

                    if company and title and url:
                        jobs.append({
                            'company': company,
                            'title': title,
                            'url': url,
                            'job_info': job_info,
                            'conditions': conditions,
                            'registration_info': registration_info,
                            'page': page_num,
                            'category': category
                        })
                except Exception as e:
                    continue

            return jobs
        except Exception as e:
            print(f"  ⚠️ 페이지 {page_num} 스크래핑 실패: {e}")
            return jobs

    def insert_jobs(self, jobs):
        """채용공고를 DB에 저장"""
        if not jobs:
            return 0

        cursor = self.connection.cursor()
        inserted = 0

        for job in jobs:
            try:
                # 중복 체크 (url과 category 조합)
                cursor.execute(
                    "SELECT id FROM jobs WHERE url = %s AND category = %s",
                    (job['url'], job['category'])
                )

                if cursor.fetchone():
                    self.stats['duplicates'] += 1
                    continue

                # 회사명 정규화 (검색용)
                company_search_key = job['company'].lower().replace(' ', '').replace('\u00A0', '')

                # INSERT
                cursor.execute("""
                    INSERT INTO jobs
                    (company, title, url, category, page, job_info, conditions,
                     registration_info, scraped_at, company_search_key)
                    VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
                """, (
                    job['company'],
                    job['title'],
                    job['url'],
                    job['category'],
                    job['page'],
                    json.dumps(job['job_info'], ensure_ascii=False),
                    json.dumps(job['conditions'], ensure_ascii=False),
                    json.dumps(job['registration_info'], ensure_ascii=False),
                    datetime.now(),
                    company_search_key
                ))
                inserted += 1
                self.stats['total_inserted'] += 1

            except Exception as e:
                self.stats['errors'] += 1
                continue

        self.connection.commit()
        cursor.close()
        return inserted

    def go_to_next_page(self):
        """다음 페이지로 이동"""
        try:
            next_btn = self.driver.find_element(
                By.XPATH,
                "//p[contains(@class, 'page3')]//a[contains(@class, 'ico next')]"
            )
            next_btn.click()
            time.sleep(2)
            return True
        except Exception:
            return False

    def scrape_category(self, category_name, category_code, max_pages=50):
        """특정 카테고리의 채용공고 스크래핑"""
        print(f"\n{'='*60}")
        print(f"🔍 {category_name} 카테고리 스크래핑 시작")
        print(f"{'='*60}\n")

        if not self.select_category(category_code):
            return

        page_num = 1
        category_scraped = 0

        while page_num <= max_pages and self.stats['total_scraped'] < 1000:
            print(f"📄 페이지 {page_num} 스크래핑 중...")

            jobs = self.scrape_page_jobs(page_num, category_code)

            if jobs:
                inserted = self.insert_jobs(jobs)
                category_scraped += len(jobs)
                self.stats['total_scraped'] += len(jobs)

                print(f"  ✅ {len(jobs)}건 수집, {inserted}건 저장 (중복: {len(jobs)-inserted}건)")
                print(f"  📊 전체 진행: {self.stats['total_scraped']}/1000건\n")
            else:
                print(f"  ⚠️ 데이터 없음\n")
                break

            # 1000건 도달 시 중단
            if self.stats['total_scraped'] >= 1000:
                print(f"\n✅ 목표 1000건 도달! 스크래핑 종료\n")
                break

            # 다음 페이지로 이동
            if not self.go_to_next_page():
                print(f"  ⏭️ 다음 페이지 없음\n")
                break

            page_num += 1
            time.sleep(1)

        print(f"\n{category_name} 카테고리: {category_scraped}건 수집 완료\n")

    def run(self):
        """메인 실행"""
        print("\n" + "="*60)
        print("🚀 Catch.co.kr 채용공고 스크래핑 시작")
        print("   목표: 최신 채용공고 1000건")
        print("="*60 + "\n")

        if not self.init_driver():
            return False

        if not self.connect_db():
            return False

        if not self.navigate_to_recruit_page():
            return False

        # IT개발 카테고리 스크래핑
        self.scrape_category("IT개발", "IT", max_pages=50)

        # 1000건 미만이면 빅데이터·AI 카테고리도 스크래핑
        if self.stats['total_scraped'] < 1000:
            # 채용 검색 페이지로 다시 이동
            self.navigate_to_recruit_page()
            remaining = 1000 - self.stats['total_scraped']
            print(f"\n📊 남은 목표: {remaining}건")
            self.scrape_category("빅데이터·AI", "BIGDATA_AI", max_pages=30)

        # 최종 통계
        print(f"\n{'='*60}")
        print("📊 최종 통계")
        print(f"{'='*60}")
        print(f"총 수집: {self.stats['total_scraped']}건")
        print(f"신규 저장: {self.stats['total_inserted']}건")
        print(f"중복 제외: {self.stats['duplicates']}건")
        print(f"오류: {self.stats['errors']}건")
        print(f"{'='*60}\n")

        self.close()
        return True


if __name__ == "__main__":
    scraper = JobScraper()
    scraper.run()
